### ISSUE TITLE
Add a way to truncate string in twig as we could under smarty

### ISSUE THREAD

**Issue by e-gaulue** [PrestaShop/PrestaShop#23456]
#### Is your feature request related to a problem?

No integrated function to truncate in twig. Was included in smarty.

#### Describe the solution you'd like

The u filter is part of the StringExtension which is not installed by default. Install it first:
`composer require twig/string-extra`

Then, use the twig/extra-bundle on Symfony projects or add the extension explicitly on the Twig environment:
```
use Twig\Extra\String\StringExtension;

$twig = new \Twig\Environment(...);
$twig->addExtension(new StringExtension());
```

#### Alternatives you've considered

Various `twig-truncate-extension` exist in composer packagist but `string-extra` is directly documented (see under).

#### Additional context

https://twig.symfony.com/doc/3.x/filters/u.html

**Comment by e-gaulue** [comment]
`string-exrta`  is the tool we need but Prestashop still support PHP 7.1 that string-extra doesn't.

So here is another alternative: dzango/twig-truncate-bundle: https://packagist.org/packages/dzango/twig-truncate-bundle

Another solution would be to create a new twig extension, but it looks it already exists.

Regards,

**Comment by hibatallahAouadni** [comment]
Hello @e-gaulue 

Thanks for the suggestion, the Developer Team will take a look.
Please stay tuned !

Ping @PrestaShop/prestashop-core-developers  , what do you think, is it doable?

**Comment by PierreRambaud** [comment]
Afaik, you're able to add it to your module and register these twig extensions.
Otherwise, I'm not sure it should be included in the Core if we don't need it :thinking:

**Comment by e-gaulue** [comment]
Here is my experience:

I used to follow Prestashop team advice regarding working around Prestashop through modules, but it was an important break. So I decided to work directly on the core. With 1.5, it was hard to maintain (with three ways merge), so for a long time, I did work directly with the core when it was really helpful and with modules otherwise. Then Prestashop has moved to git and working with the Core and keeping all up to date and clean was really easier.

According to me there are two ways featuring Prestashop:
- working for sharing (and sell) your module/themes
- working for a company that need a tool that really fit its need and play directly on the Core

Each of them has pros and cons. On my Prestashop 1.6.1.23, I had about 160 commits. Some are bug corrections, others are features, company choices... I'm migrating them presently to 1.7.7.2. I've done almost half.

My request is just a proposal and I understand your position. If you don't do it, I will have it on my side. But some complains about why developers don't make more pull request. How much are we to code directly in the Core? How much are we to consider Prestashop as a toolbox and not a final product you can play a little bit with through modules and themes? 

Presently, I'm working on grids and truncation can be sometime useful (for instance for order status), that's why I added a commit on my side and requested this feature. Playing with grids, I discovered the team creates `AbstractColumn`. Some objects relying on it have a `parent::configureOptions($resolver);` at the beginning of their own `configureOptions` and some have not. According to me, they should all have it. It makes the code homogenous, logical and so easier to read, but:
> I'm not sure it should be included in the Core if we don't need it 🤔

So, to sum it up, for those that use Prestashop as a toolbox, it would be nice to have. But usually that the same people that also know how to add it on their side.

We don't have workshop to discuss this presently, we could ask the question to slack: Should sometime Prestashop add component in the core even if not directly needed by the core? pros? cons? if yes to which extent?

I've got my opinion but I would appreciate to know the users ones, if ever it's possible to get it. 

Regards,

**Comment by matks** [comment]
> My request is just a proposal and I understand your position. If you don't do it, I will have it on my side. But some complains about why developers don't make more pull request. How much are we to code directly in the Core? How much are we to consider Prestashop as a toolbox and not a final product you can play a little bit with through modules and themes?

Hi @e-gaulue you're right to raise this point. This is indeed a very gray area.

When we accept a contribution inside the project, this means many things:
- we accept and value the contribution and its author
- we accept to maintain the new behavior for the future

This is why sometimes we reject some contributions. Each Pull Request in the project that was not planned, each suggestion for a contributor is one more item on the big pile of things we have to maintain. Also "more code" = "more bugs" 😅  this is a known effect. So it's very hard for us to say yes or no between the will to improve the project, the will to reward the hard work behind your PR (and other people PRs) and the understanding that each merged PR means we are making the quantity of code we have to maintain and keep secure bigger.

> Presently, I'm working on grids and truncation can be sometime useful (for instance for order status), that's why I added a commit on my side and requested this feature. Playing with grids, I discovered the team creates `AbstractColumn`. Some objects relying on it have a `parent::configureOptions($resolver);` at the beginning of their own `configureOptions` and some have not. According to me, they should all have it.

I completely agree. It's on our todo-list, see https://github.com/PrestaShop/PrestaShop/issues/11021 . However we are moving very slowly due to the size of the work to be done.

> We don't have workshop to discuss this presently, we could ask the question to slack: Should sometime Prestashop add component in the core even if not directly needed by the core? pros? cons? if yes to which extent?

Discussing ideas for new developer features can be done [in Slack](https://github.com/PrestaShop/open-source/tree/master/slack) in #dev-core I think it's a great idea ;)

**Comment by e-gaulue** [comment]
Hi @matks,

I know all that, and I totally agree with you. So, with this in mind, the best solution is to wait Prestashop to stop support PHP 7.1 and 7.2 and then to add `twig/string-extra` if it's not yet included in twig. It looks to be an important package, directly documented in the twig official documentation (v>=2.12.1).

Regarding the fact to have code not always useful, it needs to be discuss. 1.6 helpers could handle plenty of thing not always implemented by the core, but really useful for developers. For instance on 1.7, the `AbstractColumn` is powerful but with so few options...

For the rest, it's just I've the felling the prestashop team forget there is another way to interact with Prestashop that is to directly modify the Core code as a git branch that can easily be rebase while version are released. And it's easier to maintain than overrides because you can (most of the time) directly see the conflicts in case of new version.

Also, each time I hear calls for more developers to contribute to the project, I feel guilty. As I'm directly modifying the Core, some of my commits could be directly merge, but most of them fall in the situation you describe. And each time I try my luck, I never managed and it's time consuming. Today, I spent 3 hours on Prestashop Github to try to convince a feature is great, a bug is a bug... It's really frustrating. During this time, I did not work. I don't like to feel selfish, but often I conclue it's faster to just deal with the trouble on my side. Even reporting it, is time consuming. It is not a reproach but an observation.

Regards,

**Comment by matks** [comment]
> Hi @matks,
> 
> I know all that, and I totally agree with you. So, with this in mind, the best solution is to wait Prestashop to stop support PHP 7.1 and 7.2 and then to add `twig/string-extra` if it's not yet included in twig. It looks to be an important package, directly documented in the twig official documentation (v>=2.12.1).
> 
> Regarding the fact to have code not always useful, it needs to be discuss. 1.6 helpers could handle plenty of thing not always implemented by the core, but really useful for developers. For instance on 1.7, the `AbstractColumn` is powerful but with so few options...
> 
> For the rest, it's just I've the felling the prestashop team forget there is another way to interact with Prestashop that is to directly modify the Core code as a git branch that can easily be rebase while version are released. And it's easier to maintain than overrides because you can (most of the time) directly see the conflicts in case of new version.

I dont understand why you say we forget some people do that 🤔  ?

But what thing that is true: this git repository has, in our opinion, only one purpose: that is to generate the ZIP we build and release. This repository purpose is to store the _source code_ that aims to build a deliverable item: the PrestaShop ZIP file.

> Also, each time I hear calls for more developers to contribute to the project, I feel guilty. As I'm directly modifying the Core, some of my commits could be directly merge, but most of them fall in the situation you describe. And each time I try my luck, I never managed and it's time consuming. Today, I spent 3 hours on Prestashop Github to try to convince a feature is great, a bug is a bug... It's really frustrating. During this time, I did not work. I don't like to feel selfish, but often I conclue it's faster to just deal with the trouble on my side. Even reporting it, is time consuming. It is not a reproach but an observation.

You might be interested in [this article](https://dev.to/matks/why-should-i-contribute-to-open-source-21p) I wrote about this.

To make it short, _I do not expect you to contribute because you dont want to be selfish_. Be selfish, please. We dont want to rely on people good will because this is something that is not sustainable.

Here are some reasons (explained in the article) why you should contribute while being selfish:

- when you submit a Pull Request, you get "free review" from skilled developers and you can use it to improve your developer skills
- when a Pull Request is merged, the "maintainance burden" is passed from you to us, that means you know that this little feature you have built for yourself will now be maintained by us, not you: you have successfully made us your own subordinates!
- the efforts you put into contributing, even if it's exhausting, even if it's time-consuming, at the end of the day it's the sum of all these efforts from every contributor that make PrestaShop better

> Today, I spent 3 hours on Prestashop Github to try to convince a feature is great, a bug is a bug...

That is valuable time nonetheless. But how many Pull Requests have been submitted and merged during these 3 hours? You spent 3 hours contributing somehow to the project, and in return at the end of the day you'll be rewarded with all the work done by others. So it seems like 3 hours wasted, but actually everyday all the Pull Requests being merged in the project are developer hours that you did not pay for, and the final result will finally go into your shop. Your pocket.

So was it 3 hours wasted, or 3 hours well invested? 😄

**Comment by matks** [comment]
Looks like there was no follow-up.

@e-gaulue is the question still open ? :) adding or not this way to truncate string in twig?

I ask because maybe you discussed with other people and changed your mind :)

**Comment by e-gaulue** [comment]
To sum it up :

- the way to go is the one describe in my first post because it's a huge (must have) tool directly documented by twig. It handles truncation and a lot more. But, Prestashop still support PHP 7.1 that "string-extra" doesn't. So, it's not a solution yet. We need php: >=7.2.5.
- in the meanwhile dzango/twig-truncate-bundle works well, but I would understand it's not right integrated in Prestashop. It's in my automation scripts anyway.
- Prestashop could provide its own truncate macro, but that's a nonsense. Prestashop has moved to composer in order to use its power.

So, I would say this question could sleep until migrating to a Prestashop version with php: >=7.2.5.

Just note that string-extra allow one to call all https://symfony.com/doc/current/components/string.html from twig. So please, even if I came with a truncation need, do not hesitate to link this question with other's needs in this scope.

Regards,

**Comment by boo-code** [comment]
Done, and with the package this issue asked for. `twig/string-extra` and `twig/extra-bundle` are both in
core's `composer.json`, and `TwigExtraBundle` is registered in `app/AppKernel.php`, so the `u` filter is
available to every core and module template with no extra wiring.

Checked on a running 9.2 shop:

```
$ bin/console debug:twig | grep ' u$'
 * u
```

and rendering it:

```twig
{{ text|u.truncate(12, '...') }}          the quick...
{{ text|u.truncate(12, '...', false) }}   the quick brown...
```

(the third argument keeps whole words, which is the part `truncate` in Smarty did not give you.)

It arrived with PR #33081, "First smarty template replaced by twig", merged 2023-07-05 and released in
**9.0.0**. Worth noting how: that commit is titled "Remove twig TruncateExtension and change variable
name" - core had written its own `TruncateExtension` and deleted it in favour of `string-extra`, which is
the option this issue recommended. The objection raised in 2021 was that `string-extra` needed a PHP
version core did not support; 9.x requires 8.1, so that no longer applies.

Nothing left to build here.
